packages/telemetry/src/flight-recorder.js

```diff
@@ -0,0 +1,249 @@
+    return undefined;
+  }
+  // If the file doesn't exist, or is not large enough, create it.
+  const stbuf = await fsPromises.stat(bufferFile).catch(e => {
+    if (e.code === 'ENOENT') {
+      return undefined;
+    }
+    throw e;
+  });
+  const arenaSize = BigInt(circularBufferSize - I_ARENA_START);
+
+  const writeHeader = async () => {
+    if (stbuf && stbuf.size >= I_ARENA_START) {
+      // Header already exists.
+      return;
+    }
+
+    // Write the header.
+    const headerBuf = new Uint8Array(I_ARENA_START);
+    const header = new DataView(headerBuf.buffer);
+    header.setBigUint64(I_MAGIC, SLOG_MAGIC);
+    header.setBigUint64(I_ARENA_SIZE, arenaSize);
+    header.setBigUint64(I_CIRC_START, 0n);
+    header.setBigUint64(I_CIRC_END, 0n);
+
+    await fsPromises.mkdir(path.dirname(bufferFile), { recursive: true });
+    await fsPromises.writeFile(bufferFile, headerBuf);
+  };
+  await writeHeader();
+
+  const growFile = async () => {
+    if (stbuf && stbuf.size >= circularBufferSize) {
+      // File is big enough.
+      return;
+    }
+    // Increase the file size.;
+    await fsPromises.truncate(bufferFile, circularBufferSize);
+  };
+  await growFile();
+
+  return arenaSize;
+};
+
+export const makeMemoryMappedCircularBuffer = async ({
+  circularBufferSize = DEFAULT_CIRCULAR_BUFFER_SIZE,
+  stateDir = '/tmp',
+  circularBufferFile,
+}) => {
+  const bufferFile =
+    circularBufferFile || `${stateDir}/${DEFAULT_CIRCULAR_BUFFER_FILE}`;
+  // console.log({ circularBufferFile, bufferFile });
+
+  const newArenaSize = await initializeCircularBuffer(
+    bufferFile,
+    circularBufferSize,
+  );
+
+  /** @type {Uint8Array} */
+  const fileBuf = BufferFromFile(bufferFile).Uint8Array();
+  const header = new DataView(fileBuf.buffer, 0, I_ARENA_START);
+
+  // Detect the arena size from the header, if not initialized.
+  const hdrArenaSize = header.getBigUint64(I_ARENA_SIZE);
+  const arenaSize = newArenaSize || hdrArenaSize;
+
+  const hdrMagic = header.getBigUint64(I_MAGIC);
+  assert.equal(
+    SLOG_MAGIC,
+    hdrMagic,
+    X`${bufferFile} is not a slog buffer; wanted magic ${SLOG_MAGIC}, got ${hdrMagic}`,
+  );
+  assert.equal(
+    arenaSize,
+    hdrArenaSize,
+    X`${bufferFile} arena size mismatch; wanted ${arenaSize}, got ${hdrArenaSize}`,
+  );
+  const arena = new Uint8Array(
+    fileBuf.buffer,
+    header.byteLength,
+    Number(arenaSize),
+  );
+
+  /**
+   * @param {Uint8Array} data
+   * @param {number} [offset]
+   * @returns {IteratorResult<Uint8Array, void>}
+   */
+  const readCircBuf = (data, offset = 0) => {
+    assert(
+      offset + data.byteLength <= arenaSize,
+      X`Reading past end of circular buffer`,
+    );
+
+    // Read the data to the end of the arena.
+    let firstReadLength = data.byteLength;
+    const circStart = header.getBigUint64(I_CIRC_START);
+    const circEnd = header.getBigUint64(I_CIRC_END);
+    const readStart = (Number(circStart) + offset) % Number(arenaSize);
```

Comment:
Ah, ok, your `offset` is relative to the current trailing edge of the data. Let's make a note of that in the docstring.

That's awfully racy: most of the time (after the buffer has been filled once), `circStart` will be changing constantly, every time the writer emits a new record (assuming mostly uniform-sized records). That's fine for post-mortem reading, but it'd also be nice to have a `tail -f` mode that can correctly follow a buffer from a running process.

You've nerd-sniped me into designing a scheme with a read cursor that can tolerate concurrent writing (and can report back when it's been overrun). I don't quite have it finished yet, though, so I'll recommend landing this and then we can hack on a v2 with a more reliable read protocol.

Reply:
I'll happily turn over ownership of this, but please don't be nerd-sniped.  This implements the letter of the issue, if not the spirit, and I think it is sufficient for mainnet 1.  The `$SLOGSENDER` variable entitles anybody on the planet to do a better implementation and gain traction, without requiring our default to be all-singing all-dancing.